RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,137 @@
+# 0031 Multi Test Run Finalization
+
+# Summary
+Introduce APIs to perform data attachments reprocessing (e.g. combining, merging) after two or more independent test executions.
+
+# Motivation
+Today when Test Platform executes tests in parallel only code coverage reports are merged (attachments with uri: `datacollector://microsoft/CodeCoverage/2.0`). For other attachments processing is skipped and all attachments are returned by Test Platform.
+
+The [dotnet test](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test) command is used to execute unit tests in a given solution. The `dotnet test` command builds the solution and runs a test host application for each test project in the solution. However, currently there is no way to combine/merge attachments associated with each project execution. Coverage reports are not merged.
+
+When `Run All Tests` is performed in VS, tests for projects can be executed separately. In this case also combining/merging of data attachments is not performed. Coverage reports are not merged. `Analyze Code Coverage for All Tests` is showing coverage report for only one of the test projects.
+
+# Proposed Changes
+1. Introduce a `IDataCollectorAttachments` interface which can be implemented by Test Platform extensions and provide custom logic to combine/merge attachments. Below interface should be used only for providing logic to combine/merge information from data attachments from independent test executions. This interface should **not be** used for modifying separated data attachment. Test Platform will invoke `HandleDataCollectionAttachmentSets` only if at least 2 data attachments related to Collector (through `GetExtensionUris`) are created by independent test executions.
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection
+{
+    /// <summary>
+    /// Interface for data collectors add-ins that choose to handle generated attachments
+    /// </summary>
+    public interface IDataCollectorAttachments
```

Comment:
Could we call this an `IDataCollectorAttachmentProcessor` instead? I'd expect metadata information for attachments with `IDataCollectorAttachments`

Reply:
I totally agree. I just didn't want to modify existing interface. Can we keep existing interface and mark it as deprecated? After my changes it will not be used anywhere. Additionally we do not search for this interface in extensions.

Reply:
I agree with deprecating the old interface because there was no chance this ever worked for the intended workflow and putting out a new one with the name change.

---
In reply to: [446883700](https://github.com/microsoft/vstest-docs/pull/235#discussion_r446883700) [](ancestors = 446883700)